h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java

```diff
@@ -40,23 +38,41 @@ public String str() {
   public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
     Frame fr = stk.track(asts[1].exec(env)).getFrame();
 
-    if ((fr.numCols() != 1) || ! (fr.anyVec().isCategorical() || fr.anyVec().isString()))
-      throw new IllegalArgumentException("can only match on a single categorical/string column.");
-
     final MRTask<?> matchTask;
-    double noMatch = asts[3].exec(env).getNum();
-
+    
+    // nomatch can be NaN or number, Nan is default
+    double noMatch;
+    if (asts[3] instanceof AstNum) {
+      noMatch = asts[3].exec(env).getNum();
+    } else if (asts[3] instanceof AstId && (asts[3]).str().equals("NA")){
+      noMatch = Double.NaN;
+    } else {
+      throw new IllegalArgumentException("Expected number or 'NA'. Got: " + asts[3]);
+    }
+    
+    // start index is 1 by default
+    int startIndex;
+    if (asts[4] instanceof AstNum) {
+      startIndex = (int) asts[4].exec(env).getNum();
+      if (startIndex < 0) {
+        throw new IllegalArgumentException("Expected number >= 0. Got: " + asts[4].getClass());
+      }
+    } else {
+      throw new IllegalArgumentException("Expected number. Got: " + asts[4].getClass());
+    }
+    
     if (asts[2] instanceof AstNumList) {
-      matchTask = new NumMatchTask(((AstNumList) asts[2]).sort().expand(), noMatch);
+      matchTask = new NumMatchTask(((AstNumList) asts[2]).sort().expand(), noMatch, startIndex);
     }  else if (asts[2] instanceof AstNum) {
-      matchTask = new NumMatchTask(new double[]{asts[2].exec(env).getNum()}, noMatch);
+      matchTask = new NumMatchTask(new double[]{asts[2].exec(env).getNum()}, noMatch, startIndex);
     } else if (asts[2] instanceof AstStrList) {
       String[] values = ((AstStrList) asts[2])._strs;
-      Arrays.sort(values);
```

Comment:
I see, the values used to be sorted because we're doing a binary search below. But then it's raising an issue: binary search requires values being sorted, but then the returned positions can be wrong…
As we're now caching the positions in a hashmap, I'd suggest to get rid of both sorting and the binary search entirely. @maurever, wdyt?

Reply:
We should not sort numerical values, too. I think it is a bug. I will check and fix the numerical case as well.